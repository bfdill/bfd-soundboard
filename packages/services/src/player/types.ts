import { ChildProcess } from 'child_process'

export type DefaultPlayer =
  | 'mplayer'
  | 'afplay'
  | 'mpg123'
  | 'mpg321'
  | 'play'
  | 'omxplayer'
  | 'aplay'
  | 'cmdmp3'
export type What = string | string[]
export type PlayerOptions = {
  player: any
  players?: DefaultPlayer | DefaultPlayer[] | string[]
}
export type PlayerArgs = { [playerName: string]: string[] }
export type PlayOptions = PlayerArgs & { stdio: 'ignore' }
export type Next = Function
export interface IPlayer {
  play: (what: What, options?: PlayOptions | Next, next?: Next) => ChildProcess
}
export type createPlayer = (opts?: PlayerOptions) => IPlayer
