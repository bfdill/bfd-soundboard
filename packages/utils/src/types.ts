export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends (infer U)[]
    ? DeepPartial<U>[]
    : T[P] extends ReadonlyArray<infer U>
    ? ReadonlyArray<DeepPartial<U>>
    : T[P] extends (string | number | boolean)
    ? T[P]
    : DeepPartial<T[P]>
}
