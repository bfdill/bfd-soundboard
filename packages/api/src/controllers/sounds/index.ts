import * as KoaRouter from 'koa-router'
import { NOT_FOUND, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '@talkboy-9000/sounds'

import { IApplicationContext, ApplicationState } from '../../types'

export interface ISoundsController {
  get: (ctx: IApplicationContext) => void
}

export class SoundsController implements ISoundsController {
  private static instance: ISoundsController | undefined

  constructor(protected readonly soundService: ISoundService) {}

  get = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'SoundsController',
        method: 'get'
      }
    })
    const sounds = this.soundService.getSounds(logger)

    context.body = { sounds }

    if (sounds === null) {
      logger.warn('no sounds from service')
      context.status = NOT_FOUND
      return
    }

    logger.info({
      soundsLength: sounds.length
    })
    context.status = OK
  }

  static getInstance(): ISoundsController {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundsController(SoundService.getInstance())

    return this.instance
  }
}

export const getSoundsRouter = () =>
  new KoaRouter<ApplicationState, IApplicationContext>().get(
    '/',
    SoundsController.getInstance().get
  )
