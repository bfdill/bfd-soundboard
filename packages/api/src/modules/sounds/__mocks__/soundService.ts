import { ISoundService, Sound } from '..'

export interface IMockSoundService extends ISoundService {
  setGetBySoundId: (result: any) => void
  setGetSounds: (result: any) => void
  setIsPathValid: (result: boolean) => void
  getBySoundIdMock: () => jest.Mock
  getSoundsMock: (result: any) => jest.Mock
  IsPathValidMock: (result: boolean) => jest.Mock
}

export const getMockSoundService = (): IMockSoundService => {
  const getBySoundId = jest.fn()
  const getSounds = jest.fn()
  const isPathValid = jest.fn()

  return {
    getBySoundId,
    getSounds,
    isPathValid,
    getBySoundIdMock: getBySoundId,
    getSoundsMock: getSounds,
    IsPathValidMock: isPathValid,
    setGetBySoundId: (result: Sound | undefined) =>
      getBySoundId.mockReturnValue(result),
    setGetSounds: (result: Sound[]) => getSounds.mockReturnValue(result),
    setIsPathValid: (result: boolean) => isPathValid.mockReturnValue(result)
  } as any
}
