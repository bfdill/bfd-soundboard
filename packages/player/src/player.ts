// import * as playSound from 'play-sound'
import * as winston from 'winston'
import { createPlayer, IPlayer } from './types'
import { IPlayerService } from './player.types'

// tsd creation fail
const createPlayer: createPlayer = require('play-sound')

export class PlayerService implements IPlayerService {
  static instance: IPlayerService | undefined

  constructor(
    protected readonly player: IPlayer,
    protected readonly soundService: ISoundService
  ) {}

  playFile = (
    filename: string,
    parentLogger: winston.Logger
  ): PromiseLike<void> => {
    const logger = parentLogger.child({
      service: {
        name: 'PlayerService',
        method: 'playFile'
      }
    })

    logger.info({ filename })

    if (!this.soundService.isPathValid(filename, logger)) {
      const message = 'filename path is invalid'
      logger.error({ filename, message })
      return Promise.reject(message)
    }

    return new Promise((resolve, reject) => {
      const next = (error?: any) => {
        if (error === undefined || error === null) {
          logger.debug({
            filename,
            message: 'playback success'
          })
          resolve()
          return
        }

        logger.debug({
          error,
          filename,
          message: 'playback failure'
        })
        reject(error)
      }

      this.player.play(filename, next)
    })
  }

  static getInstance() {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerService(
      createPlayer(),
      SoundService.getInstance()
    )

    return this.instance
  }
}
