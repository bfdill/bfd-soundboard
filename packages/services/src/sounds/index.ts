import { join } from 'path'

export const PATH_TO_SOUNDS = join(process.cwd(), 'audio')
