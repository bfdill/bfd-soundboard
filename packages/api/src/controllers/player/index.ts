import * as koaRouter from 'koa-router'
import { NOT_FOUND, BAD_REQUEST, OK } from 'http-status-codes'
import { ISoundService, SoundService } from '../../modules/sounds'
import { IPlayerService, PlayerService } from '../../modules/player'
import { randomIntFromInterval } from '@bfd/utils'
import { ApplicationState, IApplicationContext } from '../../types'

export interface IPlayerController {
  playSound: (context: IApplicationContext) => PromiseLike<void>
  playRando: (context: IApplicationContext) => PromiseLike<void>
}

export class PlayerController implements IPlayerController {
  private static instance: IPlayerController | undefined

  constructor(
    protected readonly playerService: IPlayerService,
    protected readonly soundService: ISoundService
  ) {}

  playSound = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'PlayerController',
        method: 'playSound'
      }
    })
    const { soundId } = context.params
    logger.info({ soundId })

    if (typeof soundId !== 'string') {
      logger.error({
        soundId,
        message: 'soundId !== string'
      })
      context.body = { soundId }
      context.status = BAD_REQUEST
      return
    }

    const sound = this.soundService.getBySoundId(soundId, logger)

    if (sound === undefined) {
      logger.error({
        soundId,
        message: 'sound === undefined'
      })
      context.body = { soundId }
      context.status = NOT_FOUND
      return
    }

    logger.debug({
      sound,
      soundId,
      message: 'playFile'
    })

    await this.playerService.playFile(sound.filename, logger)

    context.body = { sound }
    context.status = OK
  }

  playRando = async (context: IApplicationContext) => {
    const logger = context.logger.child({
      controller: {
        name: 'PlayerController',
        method: 'playRando'
      }
    })
    logger.info('request')
    const sounds = this.soundService.getSounds(logger)

    if (sounds.length === 0) {
      logger.error({
        message: 'sounds.length === 0'
      })
      context.status = NOT_FOUND
      return
    }

    logger.silly({
      soundLength: sounds.length,
      message: 'sounds length'
    })

    const sound = sounds[randomIntFromInterval(0, sounds.length - 1)]

    logger.debug({
      sound,
      message: 'selected sound'
    })

    await this.playerService.playFile(sound.filename, logger)

    context.body = { sound }
    context.status = OK
  }

  static getInstance(): IPlayerController {
    if (this.instance !== undefined) return this.instance

    this.instance = new PlayerController(
      PlayerService.getInstance(),
      SoundService.getInstance()
    )

    return this.instance
  }
}

export const getPlayerRouter = () =>
  new koaRouter<ApplicationState, IApplicationContext>()
    .get('/rando', PlayerController.getInstance().playRando)
    .get('/:soundId', PlayerController.getInstance().playSound)
