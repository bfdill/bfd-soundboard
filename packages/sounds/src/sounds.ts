import * as sane from 'sane'
import * as winston from 'winston'
import { sync } from 'glob'
import { join, resolve, sep, basename } from 'path'
import { ISoundService, LOGGER_META } from './sounds.types'
import { Sound } from '@talkboy-9000/models'
import {
  Sounds as SoundConfig,
  ConfigurationService
} from '@talkboy-9000/configuration'

export class SoundService implements ISoundService {
  private static instance: ISoundService | undefined

  protected sounds: Sound[] = []
  protected readonly watcher: sane.Watcher

  constructor(
    public readonly config: SoundConfig,
    protected readonly saneFunction: typeof sane,
    protected readonly logger: winston.Logger
  ) {
    this.addSounds()
    this.watcher = this.createWatch()
  }

  addSounds = () => {
    this.logger.info('add sounds')

    sync(join(this.config.PathToSounds, this.config.FileGlob)).forEach(
      this.addSound
    )
  }

  addSound = (filename: string) => {
    this.logger.debug({ filename })

    this.sounds = [
      ...this.sounds,
      {
        filename,
        id: filename,
        basename: basename(filename)
      }
    ]
      .sort((a, b) => a.filename.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      filename,
      sounds: this.sounds
    })
  }

  private createWatch = (): sane.Watcher => {
    return this.saneFunction(this.config.PathToSounds, {
      glob: this.config.FileGlob
    })
      .on('add', (path: string) => {
        this.addSound(path)

        this.logger.debug({
          path,
          message: 'watch:add'
        })
      })
      .on('delete', (path: string) => {
        this.removeSound(path)

        this.logger.debug({
          path,
          message: 'watch:delete'
        })
      })
  }

  getBySoundId = (
    soundId: string,
    parentLogger: winston.Logger
  ): Sound | undefined => {
    const result = this.sounds.find(s => s.id === soundId)
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getBySoundId'
        }
      })
      .debug({
        result,
        soundId,
        message: `getBySoundId(${soundId})`
      })
    return result
  }

  getSounds = (parentLogger: winston.Logger): Sound[] => {
    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'getSounds'
        }
      })
      .silly({ sounds: this.sounds })
    return this.sounds
  }

  isPathValid = (filename: string, parentLogger: winston.Logger): boolean => {
    const absCandidate = resolve(filename) + sep
    const result =
      absCandidate.substring(0, this.config.PathToSounds.length) ===
      this.config.PathToSounds

    parentLogger
      .child({
        service: {
          name: 'SoundService',
          method: 'isPathValid'
        }
      })
      .debug({
        filename,
        result,
        message: `isPathValid(${filename})`,
        pathToSounds: this.config.PathToSounds
      })

    return result
  }

  removeSound = (filename: string) => {
    this.logger.debug(`removeSound(${filename})`)

    const soundIndex = this.sounds.findIndex(s => s.filename === filename)

    if (soundIndex === -1) return

    this.sounds = this.sounds
      .slice(0, soundIndex)
      .concat(this.sounds.slice(soundIndex + 1))
      .sort((a, b) => `${a.filename}`.localeCompare(b.filename))
      .map((v, i) => {
        v.id = `${i}`
        return v
      })

    this.logger.silly({
      message: `removeSound(${filename})`,
      sounds: this.sounds
    })
  }

  static getInstance(): ISoundService {
    if (this.instance !== undefined) return this.instance

    this.instance = new SoundService(
      ConfigurationService.getInstance().getSounds(),
      sane,
      winston.createLogger({ defaultMeta: LOGGER_META })
    )

    return this.instance
  }
}
