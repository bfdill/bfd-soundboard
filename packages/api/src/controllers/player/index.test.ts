import { PlayerController, IPlayerController } from '.'
import { IPlayerService } from '@talkboy-9000/player'
import { ISoundService } from '@talkboy-9000/sounds'
import { getJestLogger, IJestLogger } from '@talkboy-9000/winston-jest'
import { IApplicationContext } from '../../types'
import { mockApplicationState } from '../../__mocks__/applicationState'
import { Sound } from '@talkboy-9000/models'

describe('controllers -> player', () => {
  const playerService: IPlayerService = {
    playFile: jest.fn()
  } as any
  const mockGetBySoundId = jest.fn()
  const mockGetSounds = jest.fn()
  const soundService: ISoundService = {
    getBySoundId: mockGetBySoundId,
    getSounds: mockGetSounds
  } as any
  const sound: Sound = {
    basename: 'mommy shark',
    filename: 'daddy shark',
    id: 'baby_shark'
  }
  const jestLogger: IJestLogger = getJestLogger()
  const playerController: IPlayerController = new PlayerController(
    playerService,
    soundService
  )
  const contextBuilder = (soundId?: any): IApplicationContext =>
    ({
      params: { soundId },
      logger: jestLogger.logger,
      state: mockApplicationState
    } as any)

  beforeEach(() => {
    mockGetSounds.mockReturnValue([])
  })

  test('has known exports', () => {
    expect(Object.keys(require('.'))).toMatchSnapshot()
  })

  describe('playSound', () => {
    test.each([undefined, null, false, true, 1])(
      'bad request on non-string',
      async (soundId: any) => {
        expect.hasAssertions()
        const context = contextBuilder(soundId)

        await playerController.playSound(context)

        expect(context.body).toMatchSnapshot()
        expect(context.status).toMatchSnapshot()
        jestLogger.callsMatchSnapshot()
      }
    )

    test('not found on unknown song', async () => {
      expect.hasAssertions()

      const context = contextBuilder('unknown')

      await playerController.playSound(context)

      expect(context.body).toMatchSnapshot()
      expect(context.status).toMatchSnapshot()
      jestLogger.callsMatchSnapshot()
    })

    test('plays sound', async () => {
      expect.hasAssertions()

      const context = contextBuilder(sound.id)
      mockGetBySoundId.mockReturnValue(sound)

      await playerController.playSound(context)

      expect(context.body).toMatchSnapshot()
      expect(context.status).toMatchSnapshot()
      expect(playerService.playFile).toBeCalledWith(
        sound.filename,
        expect.any(Object)
      )
      jestLogger.callsMatchSnapshot()
    })
  })

  describe('playRando', () => {
    test('not found when no sounds', async () => {
      expect.hasAssertions()

      const context = contextBuilder()

      await playerController.playRando(context)

      expect(context.body).toMatchSnapshot()
      expect(context.status).toMatchSnapshot()
      jestLogger.callsMatchSnapshot()
    })

    test('plays sound', async () => {
      expect.hasAssertions()

      const context = contextBuilder(sound.id)
      mockGetBySoundId.mockReturnValue(sound)
      mockGetSounds.mockReturnValue([sound])

      await playerController.playRando(context)

      expect(context.body).toMatchSnapshot()
      expect(context.status).toMatchSnapshot()
      expect(playerService.playFile).toBeCalledWith(
        sound.filename,
        expect.any(Object)
      )
      jestLogger.callsMatchSnapshot()
    })
  })
})
