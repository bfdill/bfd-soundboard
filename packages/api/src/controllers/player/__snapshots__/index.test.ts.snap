// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`controllers -> player has known exports 1`] = `
Array [
  "PlayerController",
  "getPlayerRouter",
]
`;

exports[`controllers -> player playRando not found when no sounds 1`] = `undefined`;

exports[`controllers -> player playRando not found when no sounds 2`] = `404`;

exports[`controllers -> player playRando not found when no sounds 3`] = `
Array [
  Object {
    "controller": Object {
      "method": "playRando",
      "name": "PlayerController",
    },
    "level": "info",
    "message": "request",
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playRando\\"},\\"message\\":\\"request\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playRando not found when no sounds 4`] = `
Array [
  Object {
    "controller": Object {
      "method": "playRando",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "sounds.length === 0",
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playRando\\"},\\"message\\":\\"sounds.length === 0\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playRando plays sound 1`] = `
Object {
  "sound": Object {
    "basename": "mommy shark",
    "filename": "daddy shark",
    "id": "baby_shark",
  },
}
`;

exports[`controllers -> player playRando plays sound 2`] = `200`;

exports[`controllers -> player playRando plays sound 3`] = `
Array [
  Object {
    "controller": Object {
      "method": "playRando",
      "name": "PlayerController",
    },
    "level": "info",
    "message": "request",
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playRando\\"},\\"message\\":\\"request\\",\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playRando plays sound 4`] = `
Array [
  Object {
    "controller": Object {
      "method": "playRando",
      "name": "PlayerController",
    },
    "level": "silly",
    "message": "sounds length",
    "soundLength": 1,
    Symbol(level): "silly",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playRando\\"},\\"soundLength\\":1,\\"message\\":\\"sounds length\\",\\"level\\":\\"silly\\"}",
  },
]
`;

exports[`controllers -> player playRando plays sound 5`] = `
Array [
  Object {
    "controller": Object {
      "method": "playRando",
      "name": "PlayerController",
    },
    "level": "debug",
    "message": "selected sound",
    "sound": Object {
      "basename": "mommy shark",
      "filename": "daddy shark",
      "id": "baby_shark",
    },
    Symbol(level): "debug",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playRando\\"},\\"sound\\":{\\"basename\\":\\"mommy shark\\",\\"filename\\":\\"daddy shark\\",\\"id\\":\\"baby_shark\\"},\\"message\\":\\"selected sound\\",\\"level\\":\\"debug\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 1`] = `
Object {
  "soundId": undefined,
}
`;

exports[`controllers -> player playSound bad request on non-string 2`] = `400`;

exports[`controllers -> player playSound bad request on non-string 3`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": undefined,
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 4`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "soundId !== string",
    "soundId": undefined,
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":\\"soundId !== string\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 5`] = `
Object {
  "soundId": null,
}
`;

exports[`controllers -> player playSound bad request on non-string 6`] = `400`;

exports[`controllers -> player playSound bad request on non-string 7`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": null,
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{\\"soundId\\":null},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 8`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "soundId !== string",
    "soundId": null,
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"soundId\\":null,\\"message\\":\\"soundId !== string\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 9`] = `
Object {
  "soundId": false,
}
`;

exports[`controllers -> player playSound bad request on non-string 10`] = `400`;

exports[`controllers -> player playSound bad request on non-string 11`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": false,
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{\\"soundId\\":false},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 12`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "soundId !== string",
    "soundId": false,
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"soundId\\":false,\\"message\\":\\"soundId !== string\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 13`] = `
Object {
  "soundId": true,
}
`;

exports[`controllers -> player playSound bad request on non-string 14`] = `400`;

exports[`controllers -> player playSound bad request on non-string 15`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": true,
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{\\"soundId\\":true},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 16`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "soundId !== string",
    "soundId": true,
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"soundId\\":true,\\"message\\":\\"soundId !== string\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 17`] = `
Object {
  "soundId": 1,
}
`;

exports[`controllers -> player playSound bad request on non-string 18`] = `400`;

exports[`controllers -> player playSound bad request on non-string 19`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": 1,
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{\\"soundId\\":1},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound bad request on non-string 20`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "soundId !== string",
    "soundId": 1,
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"soundId\\":1,\\"message\\":\\"soundId !== string\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playSound not found on unknown song 1`] = `
Object {
  "soundId": "unknown",
}
`;

exports[`controllers -> player playSound not found on unknown song 2`] = `404`;

exports[`controllers -> player playSound not found on unknown song 3`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": "unknown",
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{\\"soundId\\":\\"unknown\\"},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound not found on unknown song 4`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "error",
    "message": "sound === undefined",
    "soundId": "unknown",
    Symbol(level): "error",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"soundId\\":\\"unknown\\",\\"message\\":\\"sound === undefined\\",\\"level\\":\\"error\\"}",
  },
]
`;

exports[`controllers -> player playSound plays sound 1`] = `
Object {
  "sound": Object {
    "basename": "mommy shark",
    "filename": "daddy shark",
    "id": "baby_shark",
  },
}
`;

exports[`controllers -> player playSound plays sound 2`] = `200`;

exports[`controllers -> player playSound plays sound 3`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "info",
    "message": Object {
      "soundId": "baby_shark",
    },
    Symbol(level): "info",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"message\\":{\\"soundId\\":\\"baby_shark\\"},\\"level\\":\\"info\\"}",
  },
]
`;

exports[`controllers -> player playSound plays sound 4`] = `
Array [
  Object {
    "controller": Object {
      "method": "playSound",
      "name": "PlayerController",
    },
    "level": "debug",
    "message": "playFile",
    "sound": Object {
      "basename": "mommy shark",
      "filename": "daddy shark",
      "id": "baby_shark",
    },
    "soundId": "baby_shark",
    Symbol(level): "debug",
    Symbol(message): "{\\"controller\\":{\\"name\\":\\"PlayerController\\",\\"method\\":\\"playSound\\"},\\"sound\\":{\\"basename\\":\\"mommy shark\\",\\"filename\\":\\"daddy shark\\",\\"id\\":\\"baby_shark\\"},\\"soundId\\":\\"baby_shark\\",\\"message\\":\\"playFile\\",\\"level\\":\\"debug\\"}",
  },
]
`;
